<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>tidk SI</title>
</head>

<body>
  <h1>Supplementary S2</h1>

  All of the following tests were done on an Apple M1 Macbook Pro with 8 cores and 16GB memory.

  <h2>Time complexity of <code>tidk explore</code></h2>
  <p>
    To explore the performance of <code>tidk</code>, we changed the program parameters over fasta files of different
    sizes. In this example, we used a minimum kmer length of 5, and a maximum of 30, only considering 0.1% of the length
    of the chromosome in from the chromosome ends. You can see that there is a linear increase in time taken for the
    program to run given an input fasta size, and around 20 seconds of computation time per gigabase of input DNA. Our
    fasta files were generated from copying a genome several times over to
    simulate a much larger genome. You can see the code which generated these plots at
    <a
      href="https://github.com/Euphrasiologist/tidk_paper/blob/main/src/complexity.bash">https://github.com/Euphrasiologist/tidk_paper/blob/main/src/complexity.bash</a>.
    Running over a kmer range this large is probably too much for most use cases, as most eukaryotes have much shorter
    canonical telomere repeat units (<15). </p>
      <div id="plot2"></div>
      <p>In this next plot, we lower the kmer range to one which is more widely applicable and will use far less
        compute, especially for larger genomes. Using these parameters, the program takes around half of the time to run
        as for the previous example, around 9.5 seconds per gigabase.</p>
      <div id="plot3"></div>

      <h2>Time complexity of the algorithms in <code>tidk search</code></h2>
      <p><code>tidk search</code> is at its heart a simple pattern matching algorithm, so is much faster in general than
        <code>tidk explore</code> for larger kmer ranges. Speed is around 10 seconds per gigabase of input DNA.
      </p>
      <div id="plot4"></div>
</body>

<script type="module">
  import * as Plot from "https://cdn.jsdelivr.net/npm/@observablehq/plot@0.6/+esm";

  // Time complexity for tidk explore
  const format_ticks = (d) => {
    if (d >= 1000000000) {
      return d / 1000000000 + "Gb";
    } else if (d > 1000000 && d < 1000000000) {
      return d / 1000000 + "Mb";
    } else {
      return d / 1000 + " Kb";
    }
  }

  const o_data = [
    {"Time (s)": 6.279, "Genome size (bp)": 1 * 302565711},
    {"Time (s)": 12.343, "Genome size (bp)": 2 * 302565711},
    {"Time (s)": 18.5, "Genome size (bp)": 3 * 302565711},
    {"Time (s)": 24.579, "Genome size (bp)": 4 * 302565711},
    {"Time (s)": 30.839, "Genome size (bp)": 5 * 302565711}
  ];

  const plot2 = Plot.line(o_data, {x: "Genome size (bp)", y: "Time (s)"}).plot({
    title: "Parameters: -m 5 -x 30 --distance 0.001",
    x: {tickFormat: format_ticks, grid: true},
    y: {grid: true},
    marginLeft: 50,
    marks: [
      Plot.frame(),
      Plot.link(o_data, {
        x1: 1000000000,
        x2: 1000000000,
        y1: 0,
        y2: 20.4,
        strokeDasharray: 2
      }),
      Plot.link(o_data, {
        x1: 300000000,
        x2: 1000000000,
        y1: 20.4,
        y2: 20.4,
        strokeDasharray: 2
      }),
      Plot.crosshair(o_data, {x: "Genome size (bp)", y: "Time (s)"})
    ]
  });
  const div2 = document.querySelector("#plot2");
  div2.append(plot2);

  const o_data3 = [
    {"Time (s)": 2.807, "Genome size (bp)": 1 * 302565711},
    {"Time (s)": 5.65, "Genome size (bp)": 2 * 302565711},
    {"Time (s)": 8.699, "Genome size (bp)": 3 * 302565711},
    {"Time (s)": 11.344, "Genome size (bp)": 4 * 302565711},
    {"Time (s)": 14.0, "Genome size (bp)": 5 * 302565711}
  ];

  const plot3 = Plot.line(o_data3, {x: "Genome size (bp)", y: "Time (s)"}).plot({
    title: "Parameters: -m 4 -x 15 --distance 0.001",
    x: {tickFormat: format_ticks, grid: true},
    y: {grid: true},
    marginLeft: 50,
    marks: [
      Plot.frame(),
      Plot.link(o_data3, {
        x1: 1000000000,
        x2: 1000000000,
        y1: 0,
        y2: 9.49,
        strokeDasharray: 2
      }),
      Plot.link(o_data3, {
        x1: 300000000,
        x2: 1000000000,
        y1: 9.49,
        y2: 9.49,
        strokeDasharray: 2
      }),
      Plot.crosshair(o_data3, {x: "Genome size (bp)", y: "Time (s)"})
    ]
  });

  const div3 = document.querySelector("#plot3");
  div3.append(plot3);

  // the tidk search plot
  const o_data2 = [
    {"Time (s)": 2.92, "Genome size (bp)": 1 * 302565711},
    {"Time (s)": 6.14, "Genome size (bp)": 2 * 302565711},
    {"Time (s)": 8.692, "Genome size (bp)": 3 * 302565711},
    {"Time (s)": 11.963, "Genome size (bp)": 4 * 302565711},
    {"Time (s)": 15.241, "Genome size (bp)": 5 * 302565711}
  ];

  const plot4 = Plot.line(o_data2, {x: "Genome size (bp)", y: "Time (s)"}).plot({
    title: "Parameters -s AACCT",
    x: {tickFormat: format_ticks, grid: true},
    y: {grid: true},
    marginLeft: 50,
    marks: [
      Plot.frame(),
      Plot.link(o_data2, {
        x1: 1000000000,
        x2: 1000000000,
        y1: 0,
        y2: 9.7,
        strokeDasharray: 2
      }),
      Plot.link(o_data2, {
        x1: 300000000,
        x2: 1000000000,
        y1: 9.7,
        y2: 9.7,
        strokeDasharray: 2
      }),
      Plot.crosshair(o_data2, {x: "Genome size (bp)", y: "Time (s)"})
    ]
  });

  const div4 = document.querySelector("#plot4");
  div4.append(plot4);
</script>


</html>
